### PR TITLE
community[patch]: AzureSearch async functions

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/azuresearch.py [hunk 1]
```diff
@@ -390,6 +450,63 @@ def add_embeddings(
...
+                }
+                doc.update(additional_fields)
+            data.append(doc)
+            ids.append(key)
+            # Upload data in batches
+            if len(data) == MAX_UPLOAD_BATCH_SIZE:
+                response = await self.async_client.upload_documents(documents=data)
+                # Check if all documents were successfully uploaded
+                if not all(r.succeeded for r in response):
+                    raise Exception(response)
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/azuresearch.py:494]: are they better exceptions to raise?

File: libs/community/langchain_community/vectorstores/azuresearch.py [hunk 2]
```diff
@@ -212,12 +223,19 @@ def fmt_err(x: str) -> str:
...
+        credential=credential,
+        user_agent=user_agent,
+    )
+    async_client = AsyncSearchClient(
         endpoint=endpoint,
         index_name=index_name,
         credential=credential,
         user_agent=user_agent,
     )
+    return client, async_client
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/azuresearch.py:238]: is it reasonable to only instantiate the one that's needed? are the clients stateful?

**Review by eyurtsev — APPROVED** [review]

File: libs/community/langchain_community/vectorstores/azuresearch.py (hunk 2 above)
Comment (baskaryan) [reply libs/community/langchain_community/vectorstores/azuresearch.py:238]: how do we know which is needed? you're saying for user to specify at vecstore instantiation if client should be sync or async? or to instantiate lazily?

File: libs/community/langchain_community/vectorstores/azuresearch.py [hunk 3]
```diff
@@ -681,6 +1087,44 @@ def semantic_hybrid_search_with_score(
                 if score_threshold is None or reranker_score >= score_threshold
             ]
 
+    async def asemantic_hybrid_search_with_score(
+        self,
+        query: str,
+        k: int = 4,
+        score_type: Literal["score", "reranker_score"] = "score",
```
Comment (mspronesti) [inline libs/community/langchain_community/vectorstores/azuresearch.py:1094]: I like your refactoring of the parameters to make them explicit. I guess you forgot this one though.

```suggestion
        score_type: Literal["score", "reranker_score"] = "score",
        *,
        score_threshold: Optional[float] = None
```

File: libs/community/langchain_community/vectorstores/azuresearch.py [hunk 4]
```diff
@@ -681,6 +1087,44 @@ def semantic_hybrid_search_with_score(
...
+            k (int): The number of documents to return. Default is 4.
+            score_type: Must either be "score" or "reranker_score".
+                Defaulted to "score".
+            filters: Filtering expression.
+
+        Returns:
+            List[Tuple[Document, float]]: A list of documents and their
+                corresponding scores.
+        """
+        score_threshold = kwargs.pop("score_threshold", None)
```
Comment (mspronesti) [inline libs/community/langchain_community/vectorstores/azuresearch.py:1111]:
```suggestion
```

**Pull request merged**